src/DataTypes/Serializations/SerializationObject.cpp

```diff
@@ -334,12 +374,79 @@ void SerializationObject::deserializeBinaryBulkStatePrefix(
         settings.path.pop_back();
     }
 
-    for (const auto & path : structure_state_concrete->sorted_dynamic_paths)
+    if (settings.prefixes_deserialization_thread_pool && !structure_state_concrete->sorted_dynamic_paths.empty())
     {
-        settings.path.push_back(Substream::ObjectDynamicPath);
-        settings.path.back().object_path_name = path;
-        dynamic_serialization->deserializeBinaryBulkStatePrefix(settings, object_state->dynamic_path_states[path], cache);
-        settings.path.pop_back();
+        /// Split deserialization of prefixes into several tasks and execute them in parallel inside thread pool.
+        size_t num_tasks = std::min(settings.prefixes_deserialization_thread_pool->getMaxThreads(), structure_state_concrete->sorted_dynamic_paths.size());
+        std::vector<std::shared_ptr<DeserializationTask>> tasks;
+        tasks.reserve(num_tasks);
+        /// We need to create a copy of states cache for each task, because it's not thread-safe.
+        /// We will merge them together with the original cache later.
+        std::vector<std::unique_ptr<SubstreamsDeserializeStatesCache>> caches;
+        caches.reserve(num_tasks);
+
+        /// Create an entry for each dynamic path state beforehand.
+        for (const auto & path : structure_state_concrete->sorted_dynamic_paths)
+            object_state->dynamic_path_states[path] = nullptr;
+
+        /// All threads will use the same stream getter that is not thread safe.
+        std::mutex getter_mutex;
+        auto safe_getter = [&](const SubstreamPath & path)
+        {
+            std::unique_lock lock(getter_mutex);
+            return settings.getter(path);
+        };
+
+        size_t task_size = std::max(structure_state_concrete->sorted_dynamic_paths.size() / num_tasks, 1ul);
+        for (size_t i = 0; i != num_tasks; ++i)
+        {
+            auto cache_copy = cache ? std::make_unique<SubstreamsDeserializeStatesCache>(*cache) : nullptr;
```

Comment:
Shouldn't we push cache_copy into caches immediately in case of exception in loop body? Most likely current approach is alright, since we may throw from `scheduleOrThrow`, but in that case `deserialize` won't start, I just wanted to recheck.

Reply:
Actually, I think it's better to use `trySchedule` here, if it's not scheduled for some reason (even with unlimited queue), we will execute this task in current thread later.